cpp/src/arrow/dataset/partition.cc

```diff
@@ -361,7 +412,32 @@ Result<std::string> DirectoryPartitioning::FormatValues(
     break;
   }
 
-  return fs::internal::JoinAbstractPath(std::move(segments));
+  return PartitionPathFormat{fs::internal::JoinAbstractPath(std::move(segments)), ""};
+}
+
+Result<Partitioning::PartitionPathFormat> FilenamePartitioning::FormatValues(
+    const ScalarVector& values) const {
+  std::vector<std::string> segments(static_cast<size_t>(schema_->num_fields()));
```

Comment:
This method is all the same as `DirectoryPartitioning::FormatValues` except for the last line. Can you factor out the `segments` computation?

Reply:
Yes, I thought the way to perform FilenamePartitioning should be very similar to DirectoryPartitioning. I have changed the implementation of `FormatValues`, `Inspect` and `ParseKeys` methods of these two classes to use functions present in the parent class KeyValuePartitioning which have the common operation so as to reduce code repetitions. Does this approach looks good?